File: curriculum/units/M18/code/typescript/instruction-pipeline.ts
import path from 'node:path'

export type InstructionSourceKind = 'managed' | 'user' | 'project' | 'local' | 'dynamic'
export type InstructionTrust = 'trusted' | 'approved' | 'untrusted'

export type InstructionSource = Readonly<{
  id: string
  kind: InstructionSourceKind
  filePath: string
  scopeRoot: string
  pathPrefixes?: readonly string[]
  trust: InstructionTrust
  content: string
}>

export type InstructionCatalogSnapshot = Readonly<{
  revision: number
  sources: readonly InstructionSource[]
}>

export type InstructionProjectionReport = Readonly<{
  catalogRevision: number
  discoveredCount: number
  selectedCount: number
  deduplicatedCount: number
  outOfScopeCount: number
  untrustedCount: number
  dynamicCount: number
  sourceIds: readonly string[]
}>

export type InstructionSnapshot = Readonly<{
  catalogRevision: number
  targetPath: string
  instructions: readonly Readonly<{
    id: string
    kind: InstructionSourceKind
    filePath: string
    content: string
  }>[]
  report: InstructionProjectionReport
}>

export class InstructionInvariantError extends Error {}
export class StaleInstructionRevisionError extends Error {}

export class InstructionCatalog {
  #revision = 0
  #sources: readonly InstructionSource[]

  constructor(sources: readonly InstructionSource[] = []) {
    this.#sources = validateAndFreezeSources(sources)
  }

  get revision(): number { return this.#revision }

  snapshot(): InstructionCatalogSnapshot {
    return deepFreeze({ revision: this.#revision, sources: this.#sources })
  }

  publish(expectedRevision: number, sources: readonly InstructionSource[]): InstructionCatalogSnapshot {
    if (expectedRevision !== this.#revision) {
      throw new StaleInstructionRevisionError(
        `stale instruction revision ${expectedRevision}; current=${this.#revision}`,
      )
    }
    this.#sources = validateAndFreezeSources(sources)
    this.#revision += 1
    return this.snapshot()
  }

  assertCurrent(revision: number): void {
    if (revision !== this.#revision) {
      throw new StaleInstructionRevisionError(
        `stale instruction projection ${revision}; current=${this.#revision}`,
      )
    }
  }
}

export class InstructionPipeline {
  project(
    catalog: InstructionCatalogSnapshot,
    targetPath: string,
    dynamicSources: readonly InstructionSource[] = [],
  ): InstructionSnapshot {
    const normalizedTarget = normalizeAbsolute(targetPath, 'target path')
    if (dynamicSources.some(source => source.kind !== 'dynamic')) {
      throw new InstructionInvariantError('request-only sources must use kind=dynamic')
    }
    const candidates = validateAndFreezeSources([
      ...catalog.sources,
      ...dynamicSources,
    ])
    let outOfScopeCount = 0
    let untrustedCount = 0
    const selected: InstructionSource[] = []
    for (const source of candidates) {
      if (source.trust === 'untrusted') {
        untrustedCount += 1
        continue
      }
      if (!appliesTo(source, normalizedTarget)) {
        outOfScopeCount += 1
        continue
      }
      selected.push(source)
    }

    const byPath = new Map<string, InstructionSource>()
    for (const source of selected.sort(compareSources)) {
      byPath.set(normalizeAbsolute(source.filePath, 'instruction path'), source)
    }
    const ordered = [...byPath.values()].sort(compareSources)
    return deepFreeze({
      catalogRevision: catalog.revision,
      targetPath: normalizedTarget,
      instructions: ordered.map(source => ({
        id: source.id,
        kind: source.kind,
        filePath: source.filePath,
        content: source.content,
      })),
      report: {
        catalogRevision: catalog.revision,
        discoveredCount: candidates.length,
        selectedCount: ordered.length,
        deduplicatedCount: selected.length - ordered.length,
        outOfScopeCount,
        untrustedCount,
        dynamicCount: ordered.filter(source => source.kind === 'dynamic').length,
        sourceIds: ordered.map(source => source.id),
      },
    })
  }
}

function appliesTo(source: InstructionSource, target: string): boolean {
  const root = normalizeAbsolute(source.scopeRoot, 'scope root')
  const relative = path.relative(root, target).replaceAll('\\', '/')
  if (!relative || relative === '.') return !source.pathPrefixes?.length
  if (relative.startsWith('../') || path.isAbsolute(relative)) return false
  if (!source.pathPrefixes?.length) return true
  return source.pathPrefixes.some(prefix => {
    const normalized = normalizeRelative(prefix)
    return relative === normalized || relative.startsWith(`${normalized}/`)
  })
}

function compareSources(left: InstructionSource, right: InstructionSource): number {
  return KIND_ORDER[left.kind] - KIND_ORDER[right.kind]
    || pathDepth(left.scopeRoot) - pathDepth(right.scopeRoot)
    || normalizeAbsolute(left.filePath, 'instruction path')
      .localeCompare(normalizeAbsolute(right.filePath, 'instruction path'))
    || left.id.localeCompare(right.id)
}

const KIND_ORDER: Readonly<Record<InstructionSourceKind, number>> = Object.freeze({
  managed: 0,
  user: 1,
  project: 2,
  local: 3,
  dynamic: 4,
})

function validateAndFreezeSources(sources: readonly InstructionSource[]): readonly InstructionSource[] {
  const ids = new Set<string>()
  return deepFreeze(sources.map(source => {
    requireText(source.id, 'instruction id')
    requireText(source.content, 'instruction content')
    normalizeAbsolute(source.filePath, 'instruction path')
    normalizeAbsolute(source.scopeRoot, 'scope root')
    if (ids.has(source.id)) throw new InstructionInvariantError(`duplicate instruction id: ${source.id}`)
    ids.add(source.id)
    return structuredClone(source)
  }))
}

function normalizeAbsolute(value: string, label: string): string {
  requireText(value, label)
  if (!path.isAbsolute(value)) throw new InstructionInvariantError(`${label} must be absolute`)
  return path.resolve(value).replaceAll('\\', '/').toLowerCase()
}

function normalizeRelative(value: string): string {
  requireText(value, 'path prefix')
  const normalized = value.replaceAll('\\', '/').replace(/^\.\//, '').replace(/\/$/, '')
  if (!normalized || normalized.startsWith('../') || path.isAbsolute(normalized)) {
    throw new InstructionInvariantError('path prefix must stay relative to scope root')
  }
  return normalized.toLowerCase()
}

function pathDepth(value: string): number {
  return normalizeAbsolute(value, 'scope root').split('/').filter(Boolean).length
}

function requireText(value: string, label: string): void {
  if (!value.trim()) throw new InstructionInvariantError(`${label} must not be empty`)
}

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object' && !Object.isFrozen(value)) {
    for (const nested of Object.values(value as Record<string, unknown>)) deepFreeze(nested)
    Object.freeze(value)
  }
  return value
}

File: curriculum/units/M18/code/python/instruction_pipeline.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Literal

InstructionSourceKind = Literal["managed", "user", "project", "local", "dynamic"]
InstructionTrust = Literal["trusted", "approved", "untrusted"]


@dataclass(frozen=True)
class InstructionSource:
    id: str
    kind: InstructionSourceKind
    file_path: str
    scope_root: str
    trust: InstructionTrust
    content: str
    path_prefixes: tuple[str, ...] = ()


@dataclass(frozen=True)
class InstructionCatalogSnapshot:
    revision: int
    sources: tuple[InstructionSource, ...]


@dataclass(frozen=True)
class ProjectedInstruction:
    id: str
    kind: InstructionSourceKind
    file_path: str
    content: str


@dataclass(frozen=True)
class InstructionProjectionReport:
    catalog_revision: int
    discovered_count: int
    selected_count: int
    deduplicated_count: int
    out_of_scope_count: int
    untrusted_count: int
    dynamic_count: int
    source_ids: tuple[str, ...]


@dataclass(frozen=True)
class InstructionSnapshot:
    catalog_revision: int
    target_path: str
    instructions: tuple[ProjectedInstruction, ...]
    report: InstructionProjectionReport


class InstructionInvariantError(ValueError):
    pass


class StaleInstructionRevisionError(RuntimeError):
    pass


class InstructionCatalog:
    def __init__(self, sources: tuple[InstructionSource, ...] = ()) -> None:
        self._revision = 0
        self._sources = _validate_sources(sources)

    @property
    def revision(self) -> int:
        return self._revision

    def snapshot(self) -> InstructionCatalogSnapshot:
        return InstructionCatalogSnapshot(self._revision, self._sources)

    def publish(
        self, expected_revision: int, sources: tuple[InstructionSource, ...]
    ) -> InstructionCatalogSnapshot:
        if expected_revision != self._revision:
            raise StaleInstructionRevisionError(
                f"stale instruction revision {expected_revision}; "
                f"current={self._revision}"
            )
        self._sources = _validate_sources(sources)
        self._revision += 1
        return self.snapshot()

    def assert_current(self, revision: int) -> None:
        if revision != self._revision:
            raise StaleInstructionRevisionError(
                f"stale instruction projection {revision}; "
                f"current={self._revision}"
            )


class InstructionPipeline:
    def project(
        self,
        catalog: InstructionCatalogSnapshot,
        target_path: str,
        dynamic_sources: tuple[InstructionSource, ...] = (),
    ) -> InstructionSnapshot:
        target = _normalize_absolute(target_path, "target path")
        if any(source.kind != "dynamic" for source in dynamic_sources):
            raise InstructionInvariantError(
                "request-only sources must use kind=dynamic"
            )
        candidates = _validate_sources((*catalog.sources, *dynamic_sources))
        selected: list[InstructionSource] = []
        untrusted = 0
        out_of_scope = 0
        for source in candidates:
            if source.trust == "untrusted":
                untrusted += 1
            elif not _applies_to(source, target):
                out_of_scope += 1
            else:
                selected.append(source)

        by_path: dict[str, InstructionSource] = {}
        for source in sorted(selected, key=_sort_key):
            by_path[_normalize_absolute(source.file_path, "instruction path")] = source
        ordered = tuple(sorted(by_path.values(), key=_sort_key))
        report = InstructionProjectionReport(
            catalog.revision,
            len(candidates),
            len(ordered),
            len(selected) - len(ordered),
            out_of_scope,
            untrusted,
            sum(source.kind == "dynamic" for source in ordered),
            tuple(source.id for source in ordered),
        )
        return InstructionSnapshot(
            catalog.revision,
            target,
            tuple(
                ProjectedInstruction(
                    source.id, source.kind, source.file_path, source.content
                )
                for source in ordered
            ),
            report,
        )


_KIND_ORDER = {"managed": 0, "user": 1, "project": 2, "local": 3, "dynamic": 4}


def _applies_to(source: InstructionSource, target: str) -> bool:
    root = Path(_normalize_absolute(source.scope_root, "scope root"))
    try:
        relative = Path(target).relative_to(root).as_posix().lower()
    except ValueError:
        return False
    if relative == ".":
        return not source.path_prefixes
    if not source.path_prefixes:
        return True
    return any(
        relative == prefix or relative.startswith(prefix + "/")
        for prefix in (_normalize_relative(item) for item in source.path_prefixes)
    )


def _sort_key(source: InstructionSource) -> tuple[int, int, str, str]:
    root = _normalize_absolute(source.scope_root, "scope root")
    return (
        _KIND_ORDER[source.kind],
        len(Path(root).parts),
        _normalize_absolute(source.file_path, "instruction path"),
        source.id,
    )


def _validate_sources(
    sources: tuple[InstructionSource, ...]
) -> tuple[InstructionSource, ...]:
    ids: set[str] = set()
    for source in sources:
        _require_text(source.id, "instruction id")
        _require_text(source.content, "instruction content")
        _normalize_absolute(source.file_path, "instruction path")
        _normalize_absolute(source.scope_root, "scope root")
        if source.id in ids:
            raise InstructionInvariantError(
                f"duplicate instruction id: {source.id}"
            )
        ids.add(source.id)
    return tuple(sources)


def _normalize_absolute(value: str, label: str) -> str:
    _require_text(value, label)
    path = Path(value)
    if not path.is_absolute():
        raise InstructionInvariantError(f"{label} must be absolute")
    return path.resolve().as_posix().lower()


def _normalize_relative(value: str) -> str:
    _require_text(value, "path prefix")
    normalized = value.replace("\\", "/").removeprefix("./").removesuffix("/").lower()
    if not normalized or normalized.startswith("../") or Path(normalized).is_absolute():
        raise InstructionInvariantError(
            "path prefix must stay relative to scope root"
        )
    return normalized


def _require_text(value: str, label: str) -> None:
    if not isinstance(value, str) or not value.strip():
        raise InstructionInvariantError(f"{label} must not be empty")
